go/ephemeral/team_ek.go

```diff
@@ -0,0 +1,426 @@
+			"sig":     libkb.S{Val: sig},
+			"boxes":   libkb.S{Val: string(boxesJSON)},
+		},
+	}
+	_, err = g.GetAPI().Post(apiArg)
+	return err
+}
+
+func publishNewTeamEK(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, merkleRoot libkb.MerkleRoot) (metadata keybase1.TeamEkMetadata, err error) {
+	defer g.CTrace(ctx, "publishNewTeamEK", func() error { return err })()
+
+	seed, err := newTeamEphemeralSeed()
+	if err != nil {
+		return metadata, err
+	}
+
+	// TODO enable storage
+	generation := keybase1.EkGeneration(1)
+	// storage := g.GetTeamEKBoxStorage()
+	// generation, err := storage.MaxGeneration(ctx, teamID)
+	// if err != nil {
+	// 	// Let's try to get the max from the server
+	// 	g.Log.CDebugf(ctx, "Error getting maxGeneration from storage")
+	// 	activeMetadata, err := teamEKMetadataMaybeStale(ctx, g, teamID)
+	// 	if err != nil {
+	// 		return metadata, err
+	// 	}
+	// 	if activeMetadata == nil {
+	// 		generation = 0
+	// 	} else {
+	// 		generation = activeMetadata.Generation
+	// 	}
+	// }
+	// generation++
+
+	// metadata, myTeamEKBoxed, err := signAndPublishTeamEK(ctx, g, generation, seed, merkleRoot) TODO
+	metadata, _, err = signAndPublishTeamEK(ctx, g, teamID, generation, seed, merkleRoot)
+	// TODO enable max gen retry
+	// if err != nil {
+	// 	g.Log.CDebugf(ctx, "Error posting userEK, retrying with server maxGeneration")
+	// 	// Let's retry posting with the server given max
+	// 	activeMetadata, err := teamEKMetadataMaybeStale(ctx, g, teamID)
+	// 	if err != nil {
+	// 		return metadata, err
+	// 	}
+	// 	if activeMetadata == nil {
+	// 		generation = 0
+	// 	} else {
+
+	// 		generation = activeMetadata.Generation
+	// 	}
+	// 	generation++
+	// 	metadata, myTeamEKBoxed, err = signAndPublishTeamEK(ctx, g, teamID, generation, seed, merkleRoot)
+	// 	if err != nil {
+	// 		return metadata, err
+	// 	}
+	// }
+
+	// TODO store result
+	//if myTeamEKBoxed == nil {
+	//	g.Log.CDebugf(ctx, "No box made for own userEK")
+	//} else {
+	//	err = storage.Put(ctx, teamID, generation, *myTeamEKBoxed)
+	//}
+	return metadata, err
+}
+
+func signAndPublishTeamEK(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, generation keybase1.EkGeneration, seed TeamEKSeed, merkleRoot libkb.MerkleRoot) (metadata keybase1.TeamEkMetadata, myTeamEKBoxed *keybase1.TeamEkBoxed, err error) {
+
+	dhKeypair, err := seed.DeriveDHKey()
+	if err != nil {
+		return metadata, myTeamEKBoxed, err
+	}
+
+	metadata = keybase1.TeamEkMetadata{
+		Kid:        dhKeypair.GetKID(),
+		Generation: generation,
+		HashMeta:   merkleRoot.HashMeta(),
+		// The ctime is derivable from the hash meta, by fetching the hashed
+		// root from the server, but including it saves readers a potential
+		// extra round trip.
+		Ctime: keybase1.TimeFromSeconds(merkleRoot.Ctime()),
+	}
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return metadata, myTeamEKBoxed, err
+	}
+
+	team, err := teams.Load(ctx, g, keybase1.LoadTeamArg{
```

Comment:
@mlsteele is this call considered expensive or if helpers end up calling it again is it not an issue?

Reply:
Don't use `ForceFullReload` at all. That one ditches the cache entirely.

Use `ForceRepoll` which forces a load up to date. It costs 1 api call at best (merkle leaf lookup). From my brief look at this function I think it's a good idea to set `ForceRepoll`. Since this thing is trying to post a bunch of encrypted data no matter what, with no early outs, so an extra RPC isn't going to break the bank.

Btw the costs of `ForceRepoll` costs are like this:
- If the team is not cached: Full team load. Expensive but unavoidable.
- If the team is cached but outdated: Costs an incremental load + 1 rpc. This is the case where it matters, so you want this.
- If the team is cached up to date: 1 RPC to check the merkle leaf of the team and discover we're up to date.